<a href="https://colab.research.google.com/github/Novandra7/MiniProject_DataMining/blob/main/checkpoint_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Menganalisis dan Memodelkan Data Untuk Memprediksi Spesies Bunga Iris**

#**Import Library**

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from google.colab import drive

#**BUSINESS UNDERSTANDING**

##**Business Objective**

Tujuan utama dari dataset ini adalah untuk mengembangkan model prediktif yang dapat memprediksi spesies bunga iris berdasarkan fitur-fitur morfologisnya. Hal ini dapat bermanfaat bagi para peneliti, petani, atau pecinta tumbuhan untuk mengidentifikasi spesies bunga iris dengan lebih cepat dan akurat

##**Assess Situation**

Situasi bisnis yang mendasari analisa ini adalah industri pertanian botani yang saat ini proses identifikasi dilakukan secara manual oleh ahli botani, yang memerlukan waktu dan tenaga yang cukup besar

##**Data Mining Goals**

Tujuan data mining pada dataset ini adalah mengekstraksi informasi yang berharga dan wawasan yang mendalam tentang hubungan antara fitur-fitur morfologis bunga iris dan spesiesnya

##**Project Plan**

Pada tahap awal, pemahaman data dilakukan dengan mempelajari struktur dataset iris, yang mencakup fitur-fitur seperti sepallength, sepalwidth, petallength, dan petalwidth. Setelah itu, dilanjutkan dengan eksplorasi data untuk mendapatkan pemahaman yang lebih dalam tentang fitur-fitur tersebut melalui visualisasi data. Selanjutnya, tahap pra-pemrosesan data dilakukan dengan memeriksa dan menangani missing value, serta melakukan pemrosesan data lainnya untuk mempersiapkan dataset. Pembagian data dilakukan dengan membagi dataset menjadi set pelatihan dan pengujian untuk melatih dan menguji model machine learning yang dipilih. Setelah model dipilih, dilakukan pelatihan menggunakan set pelatihan dan evaluasi kinerjanya, termasuk menggunakan teknik validasi silang jika diperlukan. Evaluasi model melibatkan perbandingan performa model yang berbeda jika digunakan lebih dari satu model, dengan opsi untuk menyetel parameter untuk meningkatkan kinerja model. Setelah model ditentukan, tahap prediksi dilakukan menggunakan model yang telah ditraining untuk membuat prediksi spesies iris berdasarkan fitur-fitur morfologisnya pada data baru atau data pengujian. Akhirnya, interpretasi hasil dilakukan untuk mengevaluasi efektivitas model dalam memprediksi spesies iris dari fitur-fitur yang diberikan.

#**DATA UNDERSTANDING**

##**Collect Initial Data**

Jelaskan darimana sumber data kalian di sini, sertakan linknya juga ya

In [ ]:
drive.mount('/content/drive')
file_path = '/content/drive/MyDrive/Colab Notebooks/iris_csv.csv'
df = pd.read_csv(file_path)

Mounted at /content/drive


##**Describe Data**

Dataset ini terdiri dari 150 sampel bunga iris yang masing-masing memiliki empat fitur: panjang sepal, lebar sepal, panjang petal, dan lebar petal. Setiap sampel bunga iris juga diberi label kelas yang menunjukkan spesies bunga iris tersebut. Deskripsi kolom-kolom dalam dataset iris adalah sebagai berikut:

1. SepalLength: Ini adalah panjang sepal (dalam cm) dari bunga iris.
2. SepalWidth: Ini adalah lebar sepal (dalam cm) dari bunga iris.
3. PetalLength: Ini adalah panjang petal (dalam cm) dari bunga iris.
4. PetalWidth: Ini adalah lebar petal (dalam cm) dari bunga iris.
5. Class: Ini adalah label kelas yang menunjukkan spesies bunga iris. Ada tiga kelas dalam dataset iris: Setosa, Versicolor, dan Virginica.

Dengan menggunakan dataset iris, Kita dapat melakukan berbagai tugas pemodelan data seperti klasifikasi, klastering, dan regresi, serta mempelajari berbagai teknik pemrosesan data dan visualisasi data.

####Informasi Dasar

In [ ]:
print("Jumlah baris:", df.shape[0])
print("Jumlah kolom:", df.shape[1])

Jumlah baris: 150
Jumlah kolom: 5


Pada dataset iris memiliki panjang kolom sebanyak 5 dan panjang baris sebanyak 150 baris

####Informasi Lanjutan

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sepallength  150 non-null    float64
 1   sepalwidth   150 non-null    float64
 2   petallength  150 non-null    float64
 3   petalwidth   150 non-null    float64
 4   class        150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


1. Kolom pertama bernama sepallenght (panjang sepal) dengan tipe data float dan tidak ada data yang kosong
1. Kolom pertama bernama sepalwidht (lebar sepal) dengan tipe data float dan tidak ada data yang kosong
1. Kolom pertama bernama petallenght (panjang petal) dengan tipe data float dan tidak ada data yang kosong
1. Kolom pertama bernama petalwidht (lebar petal) dengan tipe data float dan tidak ada data yang kosong
1. Kolom pertama bernama class (label kelas yang menunjukkan spesies bunga iris. Ada tiga kelas dalam dataset iris: Setosa, Versicolor, dan Virginica.) dengan tipe data object (string) dan tidak ada data yang kosong

####Informasi Statistik Deskriptif

In [ ]:
df.describe()

sepallength  sepalwidth  petallength  petalwidth
count   150.000000  150.000000   150.000000  150.000000
mean      5.843333    3.054000     3.758667    1.198667
std       0.828066    0.433594     1.764420    0.763161
min       4.300000    2.000000     1.000000    0.100000
25%       5.100000    2.800000     1.600000    0.300000
50%       5.800000    3.000000     4.350000    1.300000
75%       6.400000    3.300000     5.100000    1.800000
max       7.900000    4.400000     6.900000    2.500000

Sepallenght (panjang sepal)
- count = untuk kolom sepallenght memiliki **150** data yang terisi dari 150 baris data
- mean = rata rata dari kolom sepallenght adalah **5.84**
- std = standar deviasi atau persebaran data terhadap rata rata adalah **0.83**
- min = nilai terkecil dari kolom tersebut adalah **4.3**
- 25% = nilai yang membagi 25% data terendah adalah **5,1**
- 50% = nilai yang membagi 50% data terendah adalah **5,8**
- 75% = nilai yang membagi 75% data terendah adalah **6,4**
- max = nilai terbesar dari kolom tersebut adalah **7,9**

Sepalwidht (lebar sepal)
- count = untuk kolom sepalwidht memiliki **150** data yang terisi dari 150 baris data
- mean = rata rata dari kolom sepalwidht adalah **3,05**
- std = standar deviasi atau persebaran data terhadap rata rata adalah **0.43**
- min = nilai terkecil dari kolom tersebut adalah **2,0**
- 25% = nilai yang membagi 25% data terendah adalah **2,8**
- 50% = nilai yang membagi 50% data terendah adalah **3,0**
- 75% = nilai yang membagi 75% data terendah adalah **3,3**
- max = nilai terbesar dari kolom tersebut adalah **4,4**

Petallenght (panjang petal)
- count = untuk kolom petallenght memiliki **150** data yang terisi dari 150 baris data
- mean = rata rata dari kolom petallenght adalah **3,76**
- std = standar deviasi atau persebaran data terhadap rata rata adalah **1,76**
- min = nilai terkecil dari kolom tersebut adalah **1,0**
- 25% = nilai yang membagi 25% data terendah adalah **1,6**
- 50% = nilai yang membagi 50% data terendah adalah **4,35**
- 75% = nilai yang membagi 75% data terendah adalah **5,1**
- max = nilai terbesar dari kolom tersebut adalah **6,9**

Petalwidht (lebar petal)
- count = untuk kolom petalwidht memiliki **150** data yang terisi dari 150 baris data
- mean = rata rata dari kolom petalwidht adalah **1,20**
- std = standar deviasi atau persebaran data terhadap rata rata adalah **0,76**
- min = nilai terkecil dari kolom tersebut adalah **0,1**
- 25% = nilai yang membagi 25% data terendah adalah **0,3**
- 50% = nilai yang membagi 50% data terendah adalah **1,3**
- 75% = nilai yang membagi 75% data terendah adalah **1,8**
- max = nilai terbesar dari kolom tersebut adalah **2,5**



##**Exploratory Data Analysis**

In [ ]:
fig, axes = plt.subplots(4, 1, figsize=(8, 14))
sns.histplot(df['sepallength'], kde=True, ax=axes[0])
axes[0].set_title('Distribusi Panjang Sepal')
axes[0].set_xlabel('Panjang Sepal')
axes[0].set_ylabel('Frekuensi')

sns.histplot(df['sepalwidth'], kde=True, ax=axes[1])
axes[1].set_title('Distribusi Lebar Sepal')
axes[1].set_xlabel('Lebar Sepal')
axes[1].set_ylabel('Frekuensi')

sns.histplot(df['petallength'], kde=True, ax=axes[2])
axes[2].set_title('Distribusi Panjang Petal')
axes[2].set_xlabel('Panjang Petal')
axes[2].set_ylabel('Frekuensi')

sns.histplot(df['petalwidth'], kde=True, ax=axes[3])
axes[3].set_title('Distribusi Lebar Petal')
axes[3].set_xlabel('Lebar Petal')
axes[3].set_ylabel('Frekuensi')

plt.tight_layout()  # Untuk mengatur layout plot
plt.show()

- Dari plot pertama, distribusi panjang sepal terlihat cukup bervariasi, dengan beberapa puncak frekuensi yang mungkin menunjukkan keberadaan beberapa kelompok yang berbeda dalam dataset.
- Plot kedua menunjukkan distribusi lebar sepal, yang juga memiliki variasi tetapi tampaknya kurang bervariasi dibandingkan dengan panjang sepal.
- Plot ketiga menunjukkan distribusi panjang petal, yang tampaknya lebih konsisten dibandingkan dengan panjang sepal, dengan puncak frekuensi yang lebih terfokus.
- Plot keempat menunjukkan distribusi lebar petal, yang tampaknya memiliki distribusi yang paling terkonsentrasi, dengan sedikit variasi dibandingkan dengan panjang petal.
>Dari distribusi ini, kita bisa mendapatkan pemahaman yang lebih baik tentang bagaimana panjang dan lebar sepal serta panjang dan lebar petal berdistribusi dalam dataset. Hal ini dapat membantu dalam analisis lebih lanjut, seperti dalam klasifikasi atau klastering data.

In [ ]:
plt.figure(figsize=(10, 6))
sns.pairplot(df, hue='class')
plt.show()

- Dalam pairplot ini, setiap pasangan variabel diplot satu sama lain. Misalnya, jika Anda memiliki empat variabel (sepallength, sepalwidth, petallength, dan petalwidth), maka akan ada 16 subplot, termasuk yang ada di diagonal.
- Diagonal, histogram dari setiap variabel diplot, yang menunjukkan distribusi univariat dari masing-masing variabel.
- Diagonal dari setiap subplot menunjukkan distribusi variabel terkait, sedangkan plot lain menunjukkan hubungan antara dua variabel.
- Dengan menggunakan parameter hue='class', titik-titik pada scatter plot diwarnai berdasarkan kelas yang sesuai. Hal ini memungkinkan kita untuk melihat bagaimana distribusi setiap kelas terhadap pasangan variabel yang berbeda.
>Melalui pairplot ini, kita dapat melihat hubungan antara variabel-variabel yang berbeda dalam dataset, serta bagaimana kelas-kelas yang berbeda dipisahkan atau bercampur dalam ruang fitur. Ini dapat memberikan wawasan yang berharga dalam pemahaman data dan mungkin mengarah pada pemilihan fitur yang lebih baik atau pemilihan model yang sesuai untuk analisis lebih lanjut.

In [ ]:
plt.figure(figsize=(8, 6))
sns.heatmap(df.corr(), annot=True, cmap='coolwarm')
plt.title('Korelasi antara Fitur-Fitur Numerik')
plt.show()

- Heatmap ini menunjukkan korelasi antara setiap pasangan fitur numerik dalam dataset.
- Setiap sel dalam heatmap menunjukkan nilai korelasi antara dua fitur. Nilai korelasi berkisar dari -1 hingga 1, di mana -1 menunjukkan korelasi negatif sempurna, 0 menunjukkan tidak adanya korelasi, dan 1 menunjukkan korelasi positif sempurna.
- Warna dalam heatmap menggambarkan kekuatan dan arah korelasi antara dua fitur, di mana warna yang lebih terang menunjukkan korelasi yang lebih kuat (baik positif maupun negatif) dan warna yang lebih gelap menunjukkan korelasi yang lebih lemah.
- Anotasi yang ditampilkan di dalam sel-sel heatmap menunjukkan nilai korelasi yang sesuai antara dua fitur.
>Dengan melihat heatmap ini, kita dapat mengidentifikasi fitur-fitur yang memiliki korelasi yang tinggi, yang dapat menjadi penting dalam pemilihan fitur atau pemodelan lebih lanjut. Korelasi yang tinggi antara dua fitur juga dapat mengindikasikan adanya multicollinearity, yang perlu diperhatikan dalam analisis regresi.

TAMBAHKAN VISUALISASI KORELASI BESERTA PENJELASANNYA LEBIH BAIK

##**Verify Data Quality**

####Missing Values

In [ ]:
missing_values = df.isnull().sum()

print("Jumlah nilai yang hilang dalam setiap kolom:")
print(missing_values)

Jumlah nilai yang hilang dalam setiap kolom:
sepallength    0
sepalwidth     0
petallength    0
petalwidth     0
class          0
dtype: int64


Jangan lupa juga untuk berikan penjelasan di sini (baik ada nilai kosong maupun gak ada nilai kosongnya)

####Outliers Values

In [ ]:
plt.figure(figsize=(10, 6))
sns.boxplot(data=df)
plt.title('Boxplot untuk Setiap Kolom')
plt.show()

Terdapat outliner pada kolom sepalwidht. Terdapat 1 nilai outliner yang berada dibawah nilai terkecil (min) dan terdapat 3 nilai outliner diata nilai terbesar (max)

####Duplicated Values

In [ ]:
df[df.duplicated()]

sepallength  sepalwidth  petallength  petalwidth           class
34           4.9         3.1          1.5         0.1     Iris-setosa
37           4.9         3.1          1.5         0.1     Iris-setosa
142          5.8         2.7          5.1         1.9  Iris-virginica

Pada baris data ke 34, 37 dan 142 terdapat data yang duplikat

####Inconsistent Values / Noise (Optional)

Jangan lupa juga untuk berikan penjelasan di sini (OPSIONAL)